Legacy code deprecation in pass manager

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR is deprecation version of #11448 for stable/0.46.


### Details and comments

nkanazawa1989 committed:
Add deprecation warnings

nkanazawa1989 committed:
Suppress builtin pass manager deprecation warnings

nkanazawa1989 committed:
Catch deprecation for running pm

nkanazawa1989 committed:
Catch deprecation warnings in unit tests

nkanazawa1989 committed:
minor: fix test and docs

jakelishman commented on qiskit/passmanager/flow_controllers.py:
```diff
@@ -195,6 +195,10 @@ def iter_tasks(self, state: PassManagerState) -> Generator[Task, PassManagerStat
+    .. deprecated:: 0.46
+
+        This class was deprecated. This will be removed in Qiskit 1.0.
+
```
The `deprecate_func` decorator adds on this bit manually to the `__init__` functions, which is displayed in the regular class documentation.  We don't need to do the `.. deprecated:: 0.46` manually, and ideally we ought to only use the decorator, so it'll emit a `DeprecationWarning` on creation of the objects.

nkanazawa1989 replied:
Removed several ``..deprecated`` directives in [c48ef48](https://github.com/Qiskit/qiskit/pull/11449/commits/c48ef489741b8d11e6f0c5f01249e5f4263574be). I prefer to keep this particular one because `FlowController` was expected to be used as a class rather than as an instance; i.e. to add new controller to its class attribute. All class methods including `.add_flow_controller()` were already deprecated and in principle this is not necessary, but this is just to tell a user that the class itself is deprecated and they should not define new subclass.

jakelishman replied:
Sorry, I might not have been clear - the `deprecate_func` causes this exact same `.. deprecated::` directive to appear in the class documentation, if it's applied to the class's `__init__` method.  This class will have two `.. deprecated::` directives at the moment - the manual one, and the one from the decorator.

I can sort it out, though.

jakelishman replied:
I fixed this and added the `removal_timeline` argument to things in 89f2760.

jakelishman commented on qiskit/transpiler/passmanager.py:
```diff
@@ -234,8 +244,10 @@ def __add__(self, other):
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", category=DeprecationWarning)
+            flatten_tasks = list(self._flatten_tasks(self._tasks))
+            return RunningPassManager(flatten_tasks)
```
We shouldn't filter warnings within the library; it adds cost, but also, it's generally unnecessary - if we ensure that the warnings have their stack level set correctly, they'll blame Qiskit code and so not be shown to users by default.

jakelishman replied:
I removed almost all of these in 35452c5, by effectively pulling in the changes on #11448 into this branch.  The particular one highlighted here has a gross hack to make internal library use work, without modifying downstream use at all, but downstream use of `to_flow_controller` should correctly see the warnings as being blamed on Qiskit code, so they should be fine to suppress those.

nkanazawa1989 replied:
Thanks Jake. I didn't came up with this solution. I've learned a lot :)

jakelishman reviewed (commented):
It looks like the correct things have been deprecated, but I think we need to revisit how it's been done:
- we should always use the `deprecate_func` decorator to deprecate classes, so all the warnings are automatically raised correctly, and this also handles putting in `.. deprecated::` directives into the class documentation.
- we oughtn't to be using `catch_warnings` in library code.  It adds extra runtime weight, but also, if the stacklevels are set correctly, it's unnecessary; the warnings will already be filtered out, and manually catching warnings prevents users from being able to control it themselves.  We can filter the warnings that are blamed on ourselves in our own test suite, since we've already got the working version in #11448.

nkanazawa1989 committed:
Remove deprecation message from class doc

nkanazawa1989 committed:
Deprecate .passes

jakelishman committed:
Update deprecation messages

jakelishman committed:
Avoid use of `warnings.catch_warnings`

The intention here is to avoid overriding warning filters within our
code, and just eagerly migrate to the preferred paths we already know to
be identical.  The `transpiler.PassManager.to_flow_controller` hack is a
bit gross.

jakelishman committed:
Merge branch 'stable/0.46' into deprecate/passmanager_components

jakelishman reviewed (dismissed):
I pushed non-trivial changes to this and it wouldn't hurt to have another reviewer, but since everyone's swamped, we're already on top of the 0.46 deadline, and especially since 0.46 still has the old CI workflows, I'm going to move this to merge and we can modify it afterwards if anything needed changing (assuming I didn't bork the tests or lint).  Thanks for getting it sorted, Naoki.

jakelishman committed:
Add missing ConditionalController
